### PR TITLE
feat(hurl): add hurl support

### PR DESCRIPTION
**Pull request by pfeiferj**
Adds support for the hurl file extension. Hurl is a command-line rest api runner/testing engine. 

Note: .hurl files are not currently detected by vim/neovim. I have opened a PR to vim for hurl file type detection https://github.com/vim/vim/pull/12631

Grammar: https://github.com/pfeiferj/tree-sitter-hurl/
Hurl file format: https://hurl.dev/docs/hurl-file.html

### CODE REVIEW

File: lua/nvim-treesitter/parsers.lua [hunk 1]
```diff
@@ -729,6 +729,15 @@ list.http = {
   maintainers = { "@amaanq" },
 }
 
+list.hurl = {
+  install_info = {
+    url = "https://github.com/pfeiferj/tree-sitter-hurl.git",
+    branch = "main",
```
Comment (amaanq) [inline lua/nvim-treesitter/parsers.lua:735]: this can be removed, it's really about the default branch

File: queries/hurl/folds.scm [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+(body) @fold
+(response_section) @fold
+(multipart_form_data_section) @fold
+(cookies_section) @fold
+(captures_section) @fold
+(asserts_section) @fold
+(options_section) @fold
+(basic_auth_section) @fold
+(json_object) @fold
+(json_array) @fold
```
Comment (amaanq) [inline queries/hurl/folds.scm:18]:
```suggestion
[
  (comment)
  (entry)
  (request)
  (response)
  (header)
  (request_section)
  (body)
  (response_section)
  (multipart_form_data_section)
  (cookies_section)
  (captures_section)
  (asserts_section)
  (options_section)
  (basic_auth_section)
  (json_object)
  (json_array)
] @fold
```

File: queries/hurl/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,112 @@
+; highlights.scm
+
+"[QueryStringParams]" @property
+"[FormParams]" @property
+"[MultipartFormData]" @property
+"[Cookies]" @property
+"[Captures]" @property
+"[Asserts]" @property
+"[Options]" @property
+"[BasicAuth]" @property
```
Comment (amaanq) [inline queries/hurl/highlights.scm:10]:
```suggestion
[ 
  "[QueryStringParams]"
  "[FormParams]"
  "[MultipartFormData]"
  "[Cookies]"
  "[Captures]"
  "[Asserts]"
  "[Options]"
  "[BasicAuth]"
] @property
```

File: queries/hurl/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+
+(comment) @comment @spell
+
+(key_string) @property
+(json_key_string) @property
+
+(value_string) @string
+(quoted_string) @string
+(json_string) @string
```
Comment (amaanq) [inline queries/hurl/highlights.scm:19]:
```suggestion
[ 
  (value_string)
  (quoted_string)
  (json_string)
] @string
```

File: queries/hurl/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+"\\" @string.escape
+(regex_escaped_char) @string.escape
+(quoted_string_escaped_char) @string.escape
+(key_string_escaped_char) @string.escape
+(value_string_escaped_char) @string.escape
+(oneline_string_escaped_char) @string.escape
+(multiline_string_escaped_char) @string.escape
+(filename_escaped_char) @string.escape
+(json_string_escaped_char) @string.escape
```
Comment (amaanq) [inline queries/hurl/highlights.scm:31]:
```suggestion
[ 
  "\\"
  (regex_escaped_char)
  (quoted_string_escaped_char)
  (key_string_escaped_char)
  (value_string_escaped_char)
  (oneline_string_escaped_char)
  (multiline_string_escaped_char)
  (filename_escaped_char)
  (json_string_escaped_char)
] @string.escape
```

File: queries/hurl/highlights.scm [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+"regex" @function.builtin
+"variable" @function.builtin
+"duration" @function.builtin
+"sha256" @function.builtin
+"md5" @function.builtin
+"bytes" @function.builtin
+
+(filter) @attribute
+
+(version) @string.special
```
Comment (amaanq) [inline queries/hurl/highlights.scm:52]: group together with file_value and the bottom 3 special strings

File: queries/hurl/highlights.scm [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+"cookie" @function.builtin
+"body" @function.builtin
+"xpath" @function.builtin
+"jsonpath" @function.builtin
+"regex" @function.builtin
+"variable" @function.builtin
+"duration" @function.builtin
+"sha256" @function.builtin
+"md5" @function.builtin
+"bytes" @function.builtin
```
Comment (amaanq) [inline queries/hurl/highlights.scm:48]:
```suggestion
[
  "status"
  "url"
  "header"
  "cookie"
  "body"
  "xpath"
  "jsonpath"
  "regex"
  "variable"
  "duration"
  "sha256"
  "md5"
  "bytes"
] @function.builtin
```

File: queries/hurl/highlights.scm [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+"cacert" @constant.builtin
+"location" @constant.builtin
+"insecure" @constant.builtin
+"max-redirs" @constant.builtin
+"retry" @constant.builtin
+"retry-interval" @constant.builtin
+"retry-max-count" @constant.builtin
+(variable_option "variable") @constant.builtin
+"verbose" @constant.builtin
+"very-verbose" @constant.builtin
```
Comment (amaanq) [inline queries/hurl/highlights.scm:64]:
```suggestion
[
  "null"
  "cacert"
  "location"
  "insecure"
  "max-redirs"
  "retry"
  "retry-interval"
  "retry-max-count"
  "variable"
  "verbose"
  "very-verbose"
] @constant.builtin
```

File: queries/hurl/highlights.scm [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+"endsWith" @keyword.operator
+"contains" @keyword.operator
+"matches" @keyword.operator
+"exists" @keyword.operator
+"includes" @keyword.operator
+"isInteger" @keyword.operator
+"isFloat" @keyword.operator
+"isBoolean" @keyword.operator
+"isString" @keyword.operator
+"isCollection" @keyword.operator
```
Comment (amaanq) [inline queries/hurl/highlights.scm:93]:
```suggestion
[
  "not"
  "equals"
  "notEquals"
  "greaterThan"
  "greaterThanOrEquals"
  "lessThan"
  "lessThanOrEquals"
  "startsWith"
  "endsWith"
  "contains"
  "matches"
  "exists"
  "includes"
  "isInteger"
  "isFloat"
  "isBoolean"
  "isString"
  "isCollection"
] @keyword.operator

[
  "=="
  "!="
  ">"
  ">="
  "<"
  "<="
] @operator
```

File: queries/hurl/highlights.scm [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+(integer) @number
+(float) @float
+(status) @number
+(json_number) @float
```
Comment (amaanq) [inline queries/hurl/highlights.scm:98]: Hmm, should json_number be float though..

```suggestion
[
  (integer)
  (status)
] @number

[
  (float)
  (json_number)
] @float
```

File: queries/hurl/highlights.scm [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+
+":" @punctuation.delimiter
+"," @punctuation.delimiter
+
+"[" @punctuation.bracket
+"]" @punctuation.bracket
+"{" @punctuation.bracket
+"}" @punctuation.bracket
+"{{" @punctuation.special
+"}}" @punctuation.special
```
Comment (amaanq) [inline queries/hurl/highlights.scm:108]:
```suggestion
[ ":" "," ] @punctuation.delimiter

[ "[" "]" "{" "}" "{{" "}}" ] @punctuation.bracket
```

File: lua/nvim-treesitter/parsers.lua [hunk 12]
```diff
@@ -729,6 +729,15 @@ list.http = {
   maintainers = { "@amaanq" },
 }
 
+list.hurl = {
+  install_info = {
+    url = "https://github.com/pfeiferj/tree-sitter-hurl.git",
```
Comment (amaanq) [inline lua/nvim-treesitter/parsers.lua:734]: remove .git

**Review by amaanq — CHANGES_REQUESTED** [review]

**Comment by amaanq** [comment]
It looks okay mostly - just could do with better formatting of the queries, it'd be a lot more readable then. I'd also like to think some of the `@constant.builtin` keywords *could be* `@keyword`, but that's up to you to decide as I don't really know Hurl

File: queries/hurl/highlights.scm (hunk 10 above)
Comment (pfeiferj) [reply queries/hurl/highlights.scm:112]: json numbers aren't _technically_ floats but afaik they're generally used as floats. Biggest issue would be they don't actually support Inf or NaN.

**Comment by pfeiferj** [comment]
> It looks okay mostly - just could do with better formatting of the queries, it'd be a lot more readable then. I'd also like to think some of the `@constant.builtin` keywords _could be_ `@keyword`, but that's up to you to decide as I don't really know Hurl

They're really part of key-value pairs that end up getting passed as options to curl instead of being keywords that do something in hurl itself.

File: queries/hurl/highlights.scm (hunk 10 above)
Comment (amaanq) [reply queries/hurl/highlights.scm]: no worries then

**Comment by pfeiferj** [comment]
@amaanq so I have a theory as to why branch = "main" might be needed. Basically the repos that are working probably had a master branch at some point and then moved it to a main branch. So then github would resolve master to main because it knows that it's the "same" branch. I started the tree-sitter-hurl repo with a default branch of main rather than moving a branch named master to main

**Comment by pfeiferj** [comment]
> @amaanq so I have a theory as to why branch = "main" might be needed. Basically the repos that are working probably had a master branch at some point and then moved it to a main branch. So then github would resolve master to main because it knows that it's the "same" branch. I started the tree-sitter-hurl repo with a default branch of main rather than moving a branch named master to main

just confirmed, without branch = "main" it fails to install on my local machine. With branch = "main" it properly installs.

**Comment by amaanq** [comment]
No, you need an entry in lockfile.json for the commit hash, that's why it fails w/o branch

**Review by amaanq — APPROVED** [review]

**Comment by pfeiferj** [comment]
> No, you need an entry in lockfile.json for the commit hash, that's why it fails w/o branch

That makes a lot more sense, updated to include a revision and it seems to work properly.

**Comment by amaanq** [comment]
injection query is malformed

File: queries/hurl/indents.scm [hunk 13]
```diff
@@ -0,0 +1,14 @@
+; indents.scm
+
+[
+  (json_object)
+  (json_array)
+] @indent
```
Comment (lucario387) [inline queries/hurl/indents.scm:6]: `@indent.begin`

File: queries/hurl/indents.scm [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+
+[
+  "}"
+  "]"
+] @branch
```
Comment (lucario387) [inline queries/hurl/indents.scm:11]: `@indent.branch`

File: queries/hurl/indents.scm [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+
+(xml_tag) @indent
```
Comment (lucario387) [inline queries/hurl/indents.scm:13]: same here

File: queries/hurl/indents.scm [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+(xml_close_tag) @branch
```
Comment (lucario387) [inline queries/hurl/indents.scm:14]: Should there be a `@indent.end` here as well (for when editing)

File: queries/hurl/injections.scm [hunk 17]
```diff
@@ -0,0 +1,7 @@
+; injections.scm
+(json_value) @json
+(xml) @html
+
+(multiline_string
+  (multiline_string_type) @language
+  (multiline_string_content) @content @include-children @combined)
```
Comment (lucario387) [inline queries/hurl/injections.scm:7]: `@include-children` isn't a 0.8-compatible query. please remove it

Comment (lucario387) [reply queries/hurl/injections.scm:7]: If you want to use `@include-children`, then please make it a 0.9-compatible version, where you use `@injection.language`, `@injection.content`, and `(#set! injection.combined)`

Comment (pfeiferj) [reply queries/hurl/injections.scm]: It seems like it's fine to just drop include-children

File: queries/hurl/injections.scm [hunk 18]
```diff
@@ -0,0 +1,7 @@
+; injections.scm
+(json_value) @json
+(xml) @html
+
+(multiline_string
+  (multiline_string_type) @injection.language
+  (multiline_string_content) @injection.content)
```
Comment (lucario387) [inline queries/hurl/injections.scm:7]:
```suggestion
(multiline_string
  (multiline_string_type) @language
  (multiline_string_content) @content)
```

**Comment by mhoffm-aiven** [comment]
Initially I thought this was for: https://ntietz.com/blog/introducing-hurl/

**Review by lucario387 — APPROVED** [review]

**Pull request merged**